# 01/17/2022: TASK 1 CLASS SOLUTION
###  Can you predict tip amount based on guest information?

In [17]:
import numpy as np
import pandas as pd
import os
from sklearn.linear_model import LinearRegression

## import data

In [15]:
tips = pd.read_csv(os.getcwd() + r'\tips.csv')

# show top few datapoints
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

## df.describe(): analyze data for numerical columns and return statistics
* with "include" option, includes statistics for nonnumeric data

In [16]:
tips.describe(include='all')

total_bill         tip   sex smoker  day    time        size
count   244.000000  244.000000   244    244  244     244  244.000000
unique         NaN         NaN     2      2    4       2         NaN
top            NaN         NaN  Male     No  Sat  Dinner         NaN
freq           NaN         NaN   157    151   87     176         NaN
mean     19.785943    2.998279   NaN    NaN  NaN     NaN    2.569672
std       8.902412    1.383638   NaN    NaN  NaN     NaN    0.951100
min       3.070000    1.000000   NaN    NaN  NaN     NaN    1.000000
25%      13.347500    2.000000   NaN    NaN  NaN     NaN    2.000000
50%      17.795000    2.900000   NaN    NaN  NaN     NaN    2.000000
75%      24.127500    3.562500   NaN    NaN  NaN     NaN    3.000000
max      50.810000   10.000000   NaN    NaN  NaN     NaN    6.000000

### some cool conclusions:
* 244 data points were collected
* based on prices, not fine dining but not cheap either
* everyone gave a tip of at least $1 (likely based on systems used to pay)
* only open 4 days a week

## linear regression

In [19]:
# define X and y from pandas columns
X = tips.loc[:, ['total_bill', 'size']] 
print(X)
y = tips.loc[:, ['tip']]
reg = LinearRegression().fit(X, y)

     total_bill  size
0         16.99     2
1         10.34     3
2         21.01     3
3         23.68     2
4         24.59     4
..          ...   ...
239       29.03     3
240       27.18     2
241       22.67     2
242       17.82     2
243       18.78     2

[244 rows x 2 columns]


In [24]:
# y-intercept
print('Intercept: ', reg.intercept_)

# coefficients (2 because looking at both total_bill and size)
print('Coefficients: ', reg.coef_)

Intercept:  [0.66894474]
Coefficients:  [[0.09271334 0.19259779]]


The regression function is: $y = 0.6689 + 0.093\cdot\text{total_bill} + 0.192\cdot\text{size}$

In [29]:
new_total_bill = 45.25
new_size = 4

print('Predicted Tip: ', reg.predict(np.array([[new_total_bill, new_size]])))

Predicted Tip:  [[5.63461441]]


## takeaways
* look up code! you can usually find a good example in docs that you can copy and paste
* your predictive model should predict lol, u didn't have to split into train/test and focus on the accuracy on the fit
    * you should've just presented the correlation in your submission so that it can be applied generally